client/app/visualizations/chart/plotly/utils.js

```diff
@@ -0,0 +1,515 @@
+  if (useUnifiedXaxis) {
+    unifiedX = sortBy(union(...seriesList.map(s => pluck(s.data, 'x'))), identity);
+  }
+
+  return map(seriesList, (series, index) => {
+    const seriesOptions = options.seriesOptions[series.name] ||
+      { type: options.globalSeriesType };
+
+    const seriesColor = getSeriesColor(seriesOptions, index);
+
+    const plotlySeries = {
+      x: [],
+      y: [],
+      error_y: {
+        array: [],
+        color: seriesColor,
+      },
+      name: seriesOptions.name || series.name,
+      marker: { color: seriesColor },
+      insidetextfont: {
+        color: getFontColor(seriesColor),
+      },
+    };
+
+    if ((seriesOptions.yAxis === 1) && (options.series.stacking === null)) {
+      plotlySeries.yaxis = 'y2';
+    }
+
+    setType(plotlySeries, seriesOptions.type, options);
+    let data = series.data;
+    if (sortX) {
+      data = sortBy(data, 'x');
+    }
+
+    if (useUnifiedXaxis && (index === 0)) {
+      const yValues = {};
+      const eValues = {};
+
+      data.forEach((row) => {
+        yValues[row.x] = row.y;
+        if (row.yError) {
+          eValues[row.x] = row.yError;
+        }
+      });
+
+      unifiedX.forEach((x) => {
+        plotlySeries.x.push(normalizeValue(x));
+        plotlySeries.y.push(normalizeValue(yValues[x] || null));
+        if (!isUndefined(eValues[x])) {
+          plotlySeries.error_y.array.push(normalizeValue(eValues[x] || null));
+        }
+      });
+    } else {
+      data.forEach((row) => {
+        plotlySeries.x.push(normalizeValue(row.x));
+        plotlySeries.y.push(normalizeValue(row.y));
+        if (row.yError !== undefined) {
+          plotlySeries.error_y.array.push(normalizeValue(row.yError));
+        }
+      });
+    }
+    if (!plotlySeries.error_y.length) {
+      delete plotlySeries.error_y.length;
+    }
+
+    if (seriesOptions.type === 'bubble') {
+      plotlySeries.marker = {
+        size: pluck(data, 'size'),
+      };
+    } else if (seriesOptions.type === 'box') {
+      plotlySeries.boxpoints = 'outliers';
+      plotlySeries.marker = {
+        color: seriesColor,
+        size: 3,
+      };
+      if (options.showpoints) {
+        plotlySeries.boxpoints = 'all';
+        plotlySeries.jitter = 0.3;
+        plotlySeries.pointpos = -1.8;
+      }
+    }
+
+    return plotlySeries;
+  });
+}
+
+function prepareStackingData(seriesList) {
+  const xValues = union(...pluck(seriesList, 'x')).sort((a, b) => a - b);
+  seriesList.forEach((series) => {
+    series.x.sort((a, b) => a - b);
+
+    each(xValues, (value, index) => {
+      if (series.x[index] !== value) {
+        series.x.splice(index, 0, value);
+        series.y.splice(index, 0, null);
+      }
+    });
+  });
+
+  return seriesList;
+}
+
+function enableAnnotations(layout, seriesList, options) {
+  if (options.annotations) {
+    if (options.globalSeriesType === 'column') {
+      seriesList.forEach((series) => {
+        series.textposition = 'inside';
```

Comment:
> the user has three options: grouped, stacked or percent stacked. 

Then for bar chart we will have 8 options + "Disabled" instead of 4 + "Disabled" - because then there will be "Stack", "Percent stack", "Overlay", "Percent overlay"...

Maybe, we need to choose better label for that checkbox ("Percent stacking" or so), but IMHO it's not good idea to transform it into more options.

Reply:
I'm not sure I understand why the user needs to see more than only 3 options (disabled/grouped, stacked, percent)?

Reply:
Hmm. Earlier we had only tree options: "Disabled", "Stacked", "Percent stacked" - which was exactly the same as "Stacked" but with additional data processing. Now for bar charts we have more stacking variants - "Stack", "Group", "Overlay", "Relative", and each of them can have the same "percent" modifier. I decided to make it a checkbox and do not flood the dropdown with large set of values.

Reply:
About "Percent" modifier - it allows to compare relative values, in a way like pie chart. An example:

Suppose we have three series with a single value each: 100, 200, 700. For regular stacking, chart will render bar of height 100, then bar of height 200 above the first one, and the last bar - of height 700. Max value of Y-axis will be `100+200+700 = 1000`.

In relative mode, it will normalize values against their sum: `100 / 1000 * 100% = 10%`, `200 / 1000 * 100% = 20%`, `700 / 1000 * 100% = 70%`, and then will render them in the same way as described above, but max value of Y-axis will be `100%`.

For a single value probably it does not make a lot of sense, but when all series contain more that one value - there is a clear difference between regular and "percent" node - in the first mode, every bar will have different height, in the second one - every bar will have `100%`-height with values distributed in that bound - very similar to pie chart, just with bars.

Reply:
What you described is percent stacking :-) And I still don't see a difference between Stack/Relative when Percent is enabled.

I think I will just review the code itself and try to understand.

Reply:
Haha :-) There is difference too, but it will be visible if there are more than one bar, and there are bars with all positive and bars some negative values. Then in "Stacking" mode all bars will be placed above x-axis (probably negative values will not be handled properly), and in "Relative" mode bars with negative values will have absolute height "100%", but will be displayed with some offset below x-axis (depending on sum of negative values).

Reply:
I was also wondering what is the difference between all that modes, but while digging deeper into code and docs, I realized that they work differently with edge cases - like mixed positive and negative values, etc. And, as I mentioned, "Percent" modifier - it's just a modifier, it translates values, but not changes rendering algorithm.

Reply:
## Stacking

I found the documentation for the different `barmode` values [in Plotly's source](https://github.com/plotly/plotly.js/blob/5bc25b490702e5ed61265207833dbd58e8ab27f1/src/traces/bar/layout_attributes.js#L14-L29). 

Disabled is the same as `group`, because `group` is the default. `relative` seems to be the same as `stack`, but supports negative values. I tried to find a case when `stack` might be better than `relative` but couldn't think of one.

I also found [`barnorm` config value](https://github.com/plotly/plotly.js/blob/5bc25b490702e5ed61265207833dbd58e8ab27f1/src/traces/bar/layout_attributes.js#L31-L43), which allows to enable "built in" percent based rendering. Because we need to have our own calculation for area stacking, I guess we should just stick to it for bars as well.

To summarize the discussion we had so far and what I think we should do:

The user should have only 2 stacking options:

* No Stacking (we will use `group` for bars, which we can either pass explicitly now or keep as no value)
* Stacked (which for bars will use `barmode`=`relative`)

And then a checkbox to trigger "Normalize values to percentage", as you added.

These options should be only available when the user selects "Bar" or "Area" as the chart type (right? I don't see a reason to show them in other modes).

### Percent based interaction

When disabling one of the series, the 100% value should be updated - it doesn't happen at the moment.

Here's an example of what I mean:

Both series showing:

![](https://d.pr/i/qa7JVa+)

Only one enabled:

![Only One Enabled](https://d.pr/i/e267b9+)

### Migration

Moving to 2 values for stacking + toggle for percentage will require a migration. I think it's simple enough to handle it in the chart code instead of having to do a DB migration... I think I even noticed you already implemented something for it?

## Data Labels

1. Let's rename the toggle to "Show Data Labels" instead of "Annotations" and update the option name in code accordingly. Annotations have a different meaning.
2. Scatter plots (which is what we use for both lines, area and scatter) have built in support for showing data labels. You need to pass `text` to the `mode` value (and if it has a value than you use the `+` modifier, i.e. `lines+text`).